numba/npyufunc/workqueue.c

```diff
@@ -239,35 +265,97 @@ queue_state_wait(Queue *queue, int old, int repl)
 void debug_marker(void);
 void debug_marker() {};
 
+
+#ifdef _MSC_VER
+#define THREAD_LOCAL(ty) __declspec(thread) ty
+#else
+/* Non-standard C99 extension that's understood by gcc and clang */
+#define THREAD_LOCAL(ty) __thread ty
+#endif
+
+// This is the number of threads that is default, it is set on initialisation of
+// the threading backend via the launch_threads() call
+static int _INIT_NUM_THREADS = -1;
+
+// This is the per-thread thread mask, each thread can carry its own mask.
+static THREAD_LOCAL(int) _TLS_num_threads = 0;
+
+static void
+set_num_threads(int count)
+{
+    _TLS_num_threads = count;
+}
+
+static int
+get_num_threads(void)
+{
+    // This is purely to permit the implementation to survive to the point
+    // where it can exit cleanly as multiple threads cannot be used with this
+    // backend
+    if (_TLS_num_threads == 0)
+    {
+        // This is a thread that did not call launch_threads() but is still a
+        // "main" thread, probably from e.g. threading.Thread() use, it still
+        // has a TLS slot which is 0 from the lack of launch_threads() call
+        _TLS_num_threads = _INIT_NUM_THREADS;
+    }
+    return _TLS_num_threads;
+}
+
+
 // this complies to a launchable function from `add_task` like:
 // add_task(nopfn, NULL, NULL, NULL, NULL)
 // useful if you want to limit the number of threads locally
-void nopfn(void *args, void *dims, void *steps, void *data) {};
+// static void nopfn(void *args, void *dims, void *steps, void *data) {};
```

Comment:
Are all of lines 306-309 now deletable?

Reply:
I'd rather they were left in, I intermittendly end up debugging this code and am always grateful for just being able to recompile with that as something to test with.